collect_app/src/main/java/org/odk/collect/android/utilities/FormDefCache.java

```diff
@@ -0,0 +1,106 @@
+package org.odk.collect.android.utilities;
+
+import org.javarosa.core.model.FormDef;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.odk.collect.android.application.Collect;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+import io.reactivex.Completable;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+import timber.log.Timber;
+
+/** Methods for reading from and writing to the FormDef cache */
+public class FormDefCache {
+    /**
+     * Returns a RxJava Completable for serializing a FormDef and saving it in cache.
+     *
+     * @param formDef       - The FormDef to be cached.
+     * @param tempCacheFile - The temporary file into which to write the serialized FormDef.
+     * @return RxJava Completable.
+     */
+    public static Completable writeCacheAsync(FormDef formDef, File tempCacheFile) {
+        return Completable.create(emitter -> {
+            try {
+                DataOutputStream dos = new DataOutputStream(new FileOutputStream(tempCacheFile));
+                formDef.writeExternal(dos);
+                dos.flush();
+                dos.close();
+            } catch (IOException exception) {
```

Comment:
this catch block will allow any incomplete temp cache file to be renamed as the actual cache file. Hence, the catch block should be modified to something like this :

```
catch (IOException exception) {
               // delete temporary cache file
                throw exception;
            }
```

Reply:
Thanks, @shobhitagarwal1612. I’m so glad you had a look at this. I am working on this now.

Reply:
OK, here’s my proposed solution. Sorry it’s such a large change. I think it’s worth it.

https://github.com/dcbriccetti/collect/commit/1f4c4d09832debdc866fac2032a0fa9a7d1efd00
@shobhitagarwal1612, @srv-twry, your thoughts?

Reply:
@dcbriccetti Thanks for the solution. I forked your branch added some more changes to it. Please take a look at https://github.com/shobhitagarwal1612/collect/commit/fa25aa2b89db10d8d9a94d79753ee37481875ab0

Reply:
@shobhitagarwal1612 says (elsewhere) he likes the change. @srv-twry?